### PR TITLE
WIP: Added RDS Aurora service

### PR DESCRIPTION
**Pull request by jagtesh**
## Release Playbook
- [ ] Rebase against master
- [ ] Release branch ()
- [ ] Pass CI ()
- [x] Code review
- [ ] Merge into master
- [ ] Release master ()
- [ ] Pass CI ()
- [ ] Update staging rack
- [ ] Edit [Rack release record](https://github.com/convox/rack/releases) and/or update [docs](https://github.com/convox/convox.github.io)
- [ ] Publish release
- [ ] Release CLI

### CODE REVIEW

**Comment by ddollar** [comment]
Haven't tried it on a rack yet but a look over the diff gets a :+1: from me

**Comment by mwarkentin** [comment]
How does `--private` work with the `PubliclyAccessible` option in the cloudformation template?

Also, would engine version and port be things that might make sense to parameterize as well?

**Comment by jagtesh** [comment]
@mwarkentin Its the same as MySQL template for most part - except for the new key for license, and different engine, version and storage values. I haven't tested without the `--private` option for either. 

My environment isn't set up to test with development convox (besides compiling it and running tests in a docker instance). If someone can help me with that, I'd be happy to report the results.

There could be a case where different port or engine version is desired. If there's enough consensus, I can create a new PR with those changes as they'll affect MySQL and Postgres too (possibly other services too).

**Comment by mwarkentin** [comment]
Cool, thanks. :)

**Comment by nzoschke** [comment]
Awesome patch and great place to start with Aurora. So 👍  as is.

I generally prefer an iterative approach to features so I think we can merge this, then follow up with `--private`, `--engine` etc options on another PR.

**Comment by nzoschke** [comment]
I'm testing this and getting "ERROR: invalid service type: aurora". I'll try to figure out why so I can merge this.

**Comment by nzoschke** [comment]
My bad, that error is not happening. Now I'm seeing this:

ERROR: Template format error: JSON not well-formed. (line 103, column 10)

File: api/models/templates/service/aurora.tmpl [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+          "Engine": "aurora",
+          "EngineVersion": "5.6.10a",
+          "MasterUsername": { "Ref": "Username" },
+          "MasterUserPassword": { "Ref": "Password" },
+          "MultiAZ": { "Ref": "MultiAZ" },
+          "Port": "3306",
+          "PubliclyAccessible": "false",
+          "StorageType": "aurora",
+          "VPCSecurityGroups": [ { "Ref": "SecurityGroup" } ],
+          "LicenseModel": "general-public-license",
```
Comment (mwarkentin) [inline api/models/templates/service/aurora.tmpl]: @nzoschke Could be this trailing comma causing the invalid JSON error.

Comment (nzoschke) [inline api/models/templates/service/aurora.tmpl]: I think that's it. Testing...

**Comment by nzoschke** [comment]
Boom:

```
$ convox services create aurora
Creating aurora-4881 (aurora)... CREATING
```

File: api/models/templates/service/aurora.tmpl (hunk 1 above)
Comment (jagtesh) [inline api/models/templates/service/aurora.tmpl]: Sorry about that! Fixed it in the pull request as well.

**Comment by nzoschke** [comment]
Closing this and opening https://github.com/convox/rack/pull/647 for more testing / release.

Thanks @jagtesh, this is awesome.

**Comment by nzoschke** [comment]
I found another issue with this in testing:

> 1:56:54 UTC-0500	CREATE_FAILED	AWS::RDS::DBInstance	Instance	Storage size for aurora must be between 100 and 3072.

**Comment by nzoschke** [comment]
Testing if its as simple as this

```diff
diff --git a/api/models/templates/service/aurora.tmpl b/api/models/templates/service/aurora.tmpl
index 43f18b1..dc1157f 100644
--- a/api/models/templates/service/aurora.tmpl
+++ b/api/models/templates/service/aurora.tmpl
@@ -7,7 +7,7 @@
     "Parameters": {
       "AllocatedStorage": {
         "Type" : "Number",
-        "Default" : "10",
+        "Default" : "100",
         "Description" : "Allocated storage size (GB)"
       },
       "Database": {
```

**Comment by nzoschke** [comment]
That got farther but now seeing:

> 14:12:20 UTC-0500	CREATE_FAILED	AWS::RDS::DBInstance	Instance	Standalone DB Instances are not supported for the specified engine

**Comment by nzoschke** [comment]
Maybe something to do with this part in the docs:

> DBClusterIdentifier
> The identifier of an existing DB cluster that this instance will be associated with. If you specify this property, specify aurora for the Engine property and do not specify any of the following properties: AllocatedStorage, CharacterSetName, DBSecurityGroups, SourceDBInstanceIdentifier, and StorageType.

**Comment by nzoschke** [comment]
Along with the `AWS::RDS::DBCluster` resource

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html

**Comment by nzoschke** [comment]
Yeah, we jumped the gun with this, so I'm re-opening this PR.

Take a look at the example in the docs, Aurora needs a DBCluster and I think 2 DBInstances

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html

```
"RDSCluster" : {
  "Type" : "AWS::RDS::DBCluster",
  "Properties" : {
    "MasterUsername" : { "Ref" : "username" },
    "MasterUserPassword" : { "Ref" : "password" },
    "Engine" : "aurora",
    "DBSubnetGroupName" : { "Ref" : "DBSubnetGroup" },
    "DBClusterParameterGroupName" : { "Ref" : "RDSDBClusterParameterGroup" }
  }
},
"RDSDBInstance1" : {
  "Type" : "AWS::RDS::DBInstance",
  "Properties" : {
    "DBSubnetGroupName" : {
      "Ref" : "DBSubnetGroup"
    },
    "Engine" : "aurora",
    "DBClusterIdentifier" : {
      "Ref" : "RDSCluster"
    },
    "PubliclyAccessible" : "true",
    "AvailabilityZone" : { "Fn::GetAtt" : [ "Subnet1", "AvailabilityZone" ] },
    "DBInstanceClass" : "db.r3.xlarge"
  }
},
"RDSDBInstance2" : {
  "Type" : "AWS::RDS::DBInstance",
  "Properties" : {
    "DBSubnetGroupName" : {
      "Ref" : "DBSubnetGroup"
    },
    "Engine" : "aurora",
    "DBClusterIdentifier" : {
      "Ref" : "RDSCluster"
    },
    "PubliclyAccessible" : "true",
    "AvailabilityZone" : { "Fn::GetAtt" : [ "Subnet2", "AvailabilityZone" ] },
    "DBInstanceClass" : "db.r3.xlarge"
  }
}
```

**Comment by nzoschke** [comment]
Closing since there has been no response. Still, thanks for the original spike @jagtesh!

If anyone wants to continue to help with Aurora, there is a lot of good stuff in this PR and discussion.

Feel free to take this PR or to reach out in Slack and chat about it. :)

**Pull request closed without merging**